adapters/unicorn/unicorn.go

```diff
@@ -0,0 +1,251 @@
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+// MakeRequests makes the HTTP requests which should be made to fetch bids.
+func (a *adapter) MakeRequests(request *openrtb.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var extRegs openrtb_ext.ExtRegs
+	if request.Regs != nil {
+		if request.Regs.COPPA == 1 {
+			return nil, []error{&errortypes.BadInput{
+				Message: "COPPA is not supported",
+			}}
+		}
+		if err := json.Unmarshal(request.Regs.Ext, &extRegs); err == nil {
+			if extRegs.GDPR != nil && (*extRegs.GDPR == 1) {
+				return nil, []error{&errortypes.BadInput{
+					Message: "GDPR is not supported",
+				}}
+			}
+			if extRegs.USPrivacy != "" {
+				return nil, []error{&errortypes.BadInput{
+					Message: "CCPA is not supported",
+				}}
+			}
+		}
+	}
+
+	err := modifyImps(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	var modifiableSource openrtb.Source
+	if request.Source != nil {
+		modifiableSource = *request.Source
+	} else {
+		modifiableSource = openrtb.Source{}
+	}
+	modifiableSource.Ext = setSourceExt()
+	request.Source = &modifiableSource
+
+	request.Ext, err = setExt(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: getHeaders(request),
+	}
+
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func getHeaders(request *openrtb.BidRequest) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+
+	if request.Device != nil {
+		if len(request.Device.UA) > 0 {
+			headers.Add("User-Agent", request.Device.UA)
+		}
+
+		if len(request.Device.IPv6) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IPv6)
+		}
+
+		if len(request.Device.IP) > 0 {
+			headers.Add("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	return headers
+}
+
+func modifyImps(request *openrtb.BidRequest) error {
+	for i := 0; i < len(request.Imp); i++ {
+		imp := &request.Imp[i]
+
+		var ext unicornImpExt
+		err := json.Unmarshal(imp.Ext, &ext)
+
+		if err != nil {
+			return &errortypes.BadInput{
+				Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
+			}
+		}
+
+		var placementID string
+		if ext.Bidder.PlacementID != "" {
+			placementID = ext.Bidder.PlacementID
+		} else {
+			placementID, err = getStoredRequestImpID(imp)
+			if err != nil {
+				return &errortypes.BadInput{
+					Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
+				}
+			}
+		}
+
+		ext.Bidder.PlacementID = placementID
```

Comment:
The fact that there's a scenario where `ext.Bidder.PlacementID` does not get modified at all makes for an opportunity to maybe refactor the `if-else` statement in line 138 to maybe something like:
```
128           var ext unicornImpExt
129           err := json.Unmarshal(imp.Ext, &ext)
130
131           if err != nil {
132               return &errortypes.BadInput{
133                   Message: fmt.Sprintf("Error while decoding imp[%d].ext: %s", i, err),
134               }
135           }
136
137 -         var placementID string
138 -         if ext.Bidder.PlacementID != "" {
139 -             placementID = ext.Bidder.PlacementID
140 -         } else {
    +         if ext.Bidder.PlacementID == "" {
141 -             placementID, err = getStoredRequestImpID(imp)
    +             ext.Bidder.PlacementID, err = getStoredRequestImpID(imp)
142               if err != nil {
143                   return &errortypes.BadInput{
144                       Message: fmt.Sprintf("Error get StoredRequestImpID from imp[%d]: %s", i, err),
145                   }
146               }
147           }
148
149 -         ext.Bidder.PlacementID = placementID
150
151           imp.Ext, err = json.Marshal(ext)
152           if err != nil {
153               return &errortypes.BadInput{
154                   Message: fmt.Sprintf("Error while encoding imp[%d].ext: %s", i, err),
155               }
156           }
157
158           secure := int8(1)
159           imp.Secure = &secure
160 -         imp.TagID = placementID
    +         imp.TagID = ext.Bidder.PlacementID
adapters/unicorn/unicorn.go
```

Reply:
Certainly, fixed it. Thanks you.

deaf6c0